### PR TITLE
Make dropdowns optionally return event based on index, not value

### PR DESCRIPTION
**Pull request by v-rob**
This PR allows dropdown events to give the dropdown index instead of the value.  The syntax is `dropdown_index_event[<name>;<use indexing>]` where `name` is the dropdown name and `use indexing` sets the use of indexed events.  It can come before or after the dropdown; it doesn't matter.

Many people have expressed desire for this (sorcerykid and orwell are the ones I remember from off the top of my head).

## To do

This PR is Ready for Review.

## How to test

```lua
minetest.register_globalstep(function(dtime)
	local player = minetest.get_player_by_name("singleplayer")
	if player then
		local controls = player:get_player_control()
		if controls.jump then
			minetest.show_formspec("singleplayer", "formspec:test", [[
				formspec_version[4]
				size[4,5]
				dropdown[.5,.5;3,1;default;Default 1,Default 2,Default 3;2]
				dropdown[.5,2;3,1;value;Value 1,Value 2,Value 3;2;false]
				dropdown[.5,3.5;3,1;index;Index 1,Index 2,Index 3;2;true]
			]])
		end
	end
end)

minetest.register_on_player_receive_fields(function(player, formname, fields)
	if fields.default then
		minetest.chat_send_all(fields.default)
	end
	if fields.value then
		minetest.chat_send_all(fields.value)
	end
	if fields.index then
		minetest.chat_send_all(fields.index)
	end
end)
```

### CODE REVIEW

File: src/gui/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -2768,6 +2775,8 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 	m_dropdowns.clear();
 	theme_by_name.clear();
 	theme_by_type.clear();
+	field_close_on_enter.clear();
+	m_dropdown_index_event.clear();
```
Comment (Desour) [inline src/gui/guiFormSpecMenu.cpp]: Please also add this to the destructor.

**Comment by Desour** [comment]
Instead of adding a new element, you could always add to the fields table a `"<dropdown name>;index"` field, eg.:
```lua
fields = {
["drpdwn"] = "foo",
["drpdwn;index"] = "1",
}
```
This would of course be a bit weird and inconsistent.

**Comment by sorcerykid** [comment]
Thanks! But I'd prefer this to be an additional boolean parameter to the dropdown[] element in question. That way it is consistent and compatible with the implementation in ActiveFormspecs.

Here are the release notes from yesterday:
https://forum.minetest.net/viewtopic.php?p=368668#p368668

**Comment by SmallJoker** [comment]
> // TODO: remove after 2016-11-03

This is a pain. `field_close_on_enter[]` only existed for a brief moment, but this code now puts unnecessary stones in the way.

**Comment by v-rob** [comment]
So the consensus is on adding another parameter?  O.K.

**Comment by Desour** [comment]
Imo adding a new parameter to `dropdown[]` would be a bad idea.
Getting indices and not the values should be default. Having the same as for `real_coordinates` would be useful: default on for new formspec version, but also element to manually turn on/off.

**Comment by v-rob** [comment]
Hmm, that does make more sense.  In that case, I'll leave the PR as it is for now until I get more feedback.  If we do add this to the next formspec version, I'll leave that for another PR when we have more reason for version 4.

File: src/gui/guiFormSpecMenu.cpp (hunk 1 above)
Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp]: Are you sure we need to?

**Comment by sorcerykid** [comment]
> Imo adding a new parameter to `dropdown[]` would be a bad idea.

In what way is it a bad idea to simply add a new parameter. I don't see an explanation. Adding another element just to turn a feature on or off for an existing element is adding complexity to formspecs for no reason. I've encountered situations where I want one dropdown to return by-value, but another to return by-index. With a parameter that's as easy as tacking on a ";true". But with an additional element I need to keep it correctly ordered with the dropdowns.

**Comment by v-rob** [comment]
Actually... I suppose a new parameter wouldn't disclude changing the default behaviour if we ever decide to change it; for some reason I thought it did.  I apologize for that lack of foresight.  In that case, I will change it to a new parameter unless there is another objection to it.

File: src/gui/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -1323,6 +1322,16 @@ void GUIFormSpecMenu::parseDropDown(parserData* data, const std::string &element
 				<< element << "'"  << std::endl;
 }
 
+void GUIFormSpecMenu::parseDropDownIndexEvent(const std::string &element)
+{
+	std::vector<std::string> parts = split(element, ';');
+	if (parts.size() == 2 ||
+		(parts.size() > 2 && m_formspec_version > FORMSPEC_API_VERSION))
+	{
```
[the remark this answers is not in the document]
Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:1330]: Ah, the ugly double-indent continuation.  I forgot about that again.

**Comment by SmallJoker** [comment]
@sorcerykid 
> In what way is it a bad idea to simply add a new parameter. I don't see an explanation. 

What do you think about a comment like this:
```C++
// IMPORTANT! Do not add new parameters to this function until support
// for 5.3.0-dev is dropped. Old clients log a 'field_close_on_enter' warning

... ParseDropdown function contents here
```

**Comment by rubenwardy** [comment]
I don't think the close on enter parameter was ever documented in a stable release, so it should be safe to use the parameter

**Comment by v-rob** [comment]
Did dropdown ever have a close_on_enter parameter?  I thought that was only for fields and pwdfields...

File: src/gui/guiFormSpecMenu.cpp (hunk 1 above)
Comment (v-rob) [reply src/gui/guiFormSpecMenu.cpp:2806]: The other lists aren't cleared in the destructor; elements are only dropped and deleted.  I don't see how it really helps clarity or anything...

**Comment by SmallJoker** [comment]
Sorry for the false alert. This warning only appeared in text input fields.
https://github.com/minetest/minetest/blob/5.1.1/src/gui/guiFormSpecMenu.cpp#L1006 <-- none for dropdown. In this case it would be saner to append a new param.

**Comment by v-rob** [comment]
OK, I've changed it to be a new parameter and bumped the formspec version to 4 accordingly.  If the version is 4, the event is automatically set to index.  I don't know if that is desired (and it is a little confusing to be honest in the docs), but it's very easily revertable.

**Comment by v-rob** [comment]
OK, I've changed it to be a new parameter and bumped the formspec version to 4 accordingly.  The default is still set to value because it would be confusing in the docs otherwise and its really easy to change now that it's a parameter.

File: doc/lua_api.txt [hunk 3]
```diff
@@ -4485,7 +4493,7 @@ Call these functions only at load time!
         * `button` and variants: If pressed, contains the user-facing button
           text as value. If not pressed, is `nil`
         * `field`, `textarea` and variants: Text in the field
-        * `dropdown`: Text of selected item
+        * `dropdown`: Text of selected item or index if `dropdown_index_event` is true
```
Comment (SmallJoker) [inline doc/lua_api.txt:4496]: Outdated.
```suggestion
        * `dropdown`: Value depending on definition
```

File: doc/lua_api.txt [hunk 4]
```diff
@@ -4485,7 +4493,7 @@ Call these functions only at load time!
         * `button` and variants: If pressed, contains the user-facing button
           text as value. If not pressed, is `nil`
         * `field`, `textarea` and variants: Text in the field
-        * `dropdown`: Text of selected item
+        * `dropdown`: Value depending on definition
```
Comment (rubenwardy) [inline doc/lua_api.txt:4496]:
```suggestion
        * `dropdown`: Either the id or text, depending on the `index event` dropdown argument.
```

**Pull request merged**